### PR TITLE
[Php83] Adds rule for adding Override attribute

### PR DESCRIPTION
**Pull request by peterfox**
The [Override](https://wiki.php.net/rfc/marking_overriden_methods) attribute is being introduced in PHP 8.3

This rule will detect when a class is overriding a method from a parent.

### CODE REVIEW

**Comment by peterfox** [comment]
@TomasVotruba @staabm I've written this new rule but finding a strange issue with the test suite. Namely:

```shell
./vendor/bin/phpunit rules-tests/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector
```

Will complete the tests and run them fine, but when I run the full suite it's as if Rector isn't picking up the rule. It literally won't even run the nodeTypes() method of the rule.

Is it because it's a PHP83 rule? Is there something missing that needs to be added?

File: src/ValueObject/PhpVersionFeature.php [hunk 1]
```diff
@@ -643,4 +643,10 @@ final class PhpVersionFeature
      * @var int
      */
     public const SENSITIVE_PARAMETER_ATTRIBUTE = PhpVersion::PHP_82;
+
+    /**
+     * @see https://wiki.php.net/rfc/marking_overriden_methods
+     * @var int
+     */
+    public const OVERIDE_ATTRIBUTE = PhpVersion::PHP_83;
```
Comment (staabm) [inline src/ValueObject/PhpVersionFeature.php]:
```suggestion
    public const OVERRIDE_ATTRIBUTE = PhpVersion::PHP_83;
```

**Review by staabm — COMMENTED** [review]
> Is it because it's a PHP83 rule? Is there something missing that needs to be added?

I didn't see something obvious wrong. maybe @samsonasik knows more.

File: rules-tests/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector/config/configured_rule.php [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+use Rector\Config\RectorConfig;
+use Rector\Php83\Rector\ClassMethod\AddOverrideAttributeToOverriddenMethodsRector;
+
+return static function (RectorConfig $rectorConfig): void {
+    $rectorConfig->rule(AddOverrideAttributeToOverriddenMethodsRector::class);
```
Comment (samsonasik) [inline rules-tests/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector/config/configured_rule.php:10]: you can add php 8.3 `$rectorConfig->phpVersion()` here.

**Comment by peterfox** [comment]
@samsonasik this appears to have worked so thank you. My only issue now is adding PHP83 to src has triggered an issue with the [Doctrine package's PHPStan](https://github.com/rectorphp/rector-src/actions/runs/6532107022/job/17734661249?pr=5170). What do you think the best way is to resolve this? Creating a PR/branch for the package?

File: rules-tests/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector/Fixture/apply_attribute_to_overridden_methods.php.inc [hunk 3]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+namespace Rector\Tests\Php83\Rector\ClassMethod\AddOverrideAttributeToOverriddenMethodsRector\Fixture;
+
+use Rector\Tests\Php83\Rector\ClassMethod\AddOverrideAttributeToOverriddenMethodsRector\Source\ExampleParentClass;
+
+class ChildClass extends ExampleParentClass
```
Comment (samsonasik) [inline rules-tests/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector/Fixture/apply_attribute_to_overridden_methods.php.inc:7]: no parent class fixture should be added:

```php
class SkipNoParentClass
{
    public function foo()
    {
    }
}
```

File: rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+        // Fetch the parent class reflection
+        $parentClassReflection = $this->reflectionProvider->getClass((string) $node->extends);
+
+        foreach ($node->getMethods() as $method) {
+            if ($method->name->toString() === '__construct') {
+                continue;
+            }
+            // Private methods should be ignored
+            if ($parentClassReflection->hasMethod($method->name->toString())) {
```
Comment (samsonasik) [inline rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php]: should be use native method exists check

```suggestion
            if ($parentClassReflection->hasNativeMethod($method->name->toString())) {
```

File: rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                // ignore if it is a private method on the parent
+                $parentMethod = $parentClassReflection->getMethod($method->name->toString(), $scope);
```
Comment (samsonasik) [inline rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php]: should be use native method exists check

```suggestion
                $parentMethod = $parentClassReflection->getNativeMethod($method->name->toString(), $scope);
```

File: rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+    private function shouldSkipClass(Class_ $class): bool
+    {
+        if ($this->classAnalyzer->isAnonymousClass($class)) {
+            return true;
+        }
+
+        return $class->extends instanceof FullyQualified && ! $this->reflectionProvider->hasClass(
+            $class->extends->toString()
+        );
```
Comment (samsonasik) [inline rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php]: should be more readable with check extends exists first:

```suggestion
        if (! $class->extends instanceof FullyQualified) {
                return true;
        }

        return ! $this->reflectionProvider->hasClass($class->extends->toString());
```

File: rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                if (! $this->nodeNameResolver->isName($attribute->name, 'Override')) {
+                    continue;
+                }
+
+                return true;
+            }
+        }
+
+        return false;
+    }
```
Comment (samsonasik) [inline rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php]: there is already service for it:

```
if ($this->phpAttributeAnalyzer->hasPhpAttribute($node, 'Override')) {
```

File: rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    public function refactor(Node $node): ?Node
+    {
+        // Detect if class extends a parent class
+        if ($this->shouldSkipClass($node)) {
+            return null;
+        }
+
+        // Fetch the parent class reflection
+        $parentClassReflection = $this->reflectionProvider->getClass((string) $node->extends);
+
```
Comment (samsonasik) [inline rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php:97]: when no change on loop, the flag is needed, eg: 

```php
$hasChanged = false;

foreach (...

   // new attribitute?
   $hasChanged = true;

}
// before return, check
if (! $hasChanged) {
   return null;
}

return $node;

File: rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        foreach ($node->getMethods() as $method) {
```
Comment (samsonasik) [inline rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php:100]:
```suggestion
        $hasChanged = false;
        foreach ($node->getMethods() as $method) {
```

File: rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+                // ignore if it is a private method on the parent
+                $parentMethod = $parentClassReflection->getNativeMethod($method->name->toString());
+                if ($parentMethod->isPrivate()) {
+                    continue;
+                }
+                // ignore if it already uses the attribute
+                if ($this->phpAttributeAnalyzer->hasPhpAttribute($method, 'Override')) {
+                    continue;
+                }
+                $method->attrGroups[] = new AttributeGroup([new Attribute(new FullyQualified('Override'))]);
```
Comment (samsonasik) [inline rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php:115]:
```suggestion
                $method->attrGroups[] = new AttributeGroup([new Attribute(new FullyQualified('Override'))]);
                $hasChanged = true;
```

File: rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+            }
+        }
+
+        return $node;
```
Comment (samsonasik) [inline rules/Php83/Rector/ClassMethod/AddOverrideAttributeToOverriddenMethodsRector.php:124]:
```suggestion
        if (! $hasChanged) {
             return null;
        }
        
        return $node;
```

**Review by samsonasik — APPROVED** [review]

**Comment by TomasVotruba** [comment]
Thank you @peterfox for this new rule :clap:

**Pull request merged**